### PR TITLE
Add email link to set password. See issue #1205.

### LINKED ISSUE #1205: Customer cannot see password in email when it is set in backend
When a customer is created in the backend, the password is not shown in the email [password_new.html](https://github.com/OpenMage/magento-lts/blob/1.9.4.x/app/locale/en_US/template/email/password_new.html)

### Preconditions (*)

1. OpenMage LTS 19.4.6, with [commit 32d9d5](https://github.com/openmage/magento-lts/commit/32d9d52a1dd5daf50513b3ababe22afc8a6b908d)
2. 

### Steps to reproduce (*)

1. Backend > Customers > Manage Customers > click _Add New Customer_ to create a customer
2. Before saving the customer, click checkbox _Send Auto-Generated Password_, then save the customer.

### Expected result (*)

1. The new passoword email has the password in plaintext for the customer to login.
2.

### Actual result (*)

1. Customer cannot see the password generated by the system
> Password: (the password you set when creating your account) 
2.

### PR DESCRIPTION
**Pull request by kiatng**
Trying to rebase PR #1283 to the `main` branch. 

### Fixed Issues (if relevant)
1. Fixes #1205

### Description (*)
When a customer account is created in backed, the password should not be emailed in plaintext. This PR add a new email template, which is modifed from the welcome email, by adding a link to set password.

image

 For existing account, the email template used is the same as the forgot-password email.

This PR does not change existing feature on emailing plaintext password from backend, but it'll show a warning:
image

### Manual testing scenarios (*)

There are 2 scenarios:

#### Creating New Customer
1. In backend, create a new customer with an email address which you own.
2. When you receive the email, click on the link to set the password.
3. Login with the new password.

#### Existing Customer
1. In backend customer page, go to the customer you created above.
2. Click on _Email Link to Set Password_.
3. When you receive the email, click on the link to reset the password.
4. Login with the new password.

### Questions or comments

The landing page of the link is `customer/account/changeforgotten` with page title **RESET PASSWORD**. See [screenshot](https://github.com/OpenMage/magento-lts/issues/1205#issuecomment-690806216). Does it need to be changed?

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Newpass.php [hunk 1]
```diff
@@ -35,6 +35,7 @@ public function render(Varien_Data_Form_Element_Abstract $element)
         if ($element->getNote()) {
             $html .= '<p class="note"><span>' . $element->getNote() . '</span></p>';
         }
+        $html .= '<p id="email-passowrd-warning" style="display:none;" class="note"><span>' . Mage::helper('customer')->__('Warning: email cotains password in plaintext will be sent.') . '</span></p>';
```
Comment (fballiano) [inline app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Newpass.php:38]:
```suggestion
        $html .= '<p id="email-passowrd-warning" style="display:none;" class="note"><span>' . Mage::helper('customer')->__('Warning: email containing password in plaintext will be sent.') . '</span></p>';
```

File: app/locale/en_US/Mage_Customer.csv [hunk 2]
```diff
@@ -451,6 +457,7 @@
 "View Order","View Order"
 "Visitor","Visitor"
 "Visitors Only","Visitors Only"
+"Warning: email cotains password in plaintext will be sent.","Warning: email cotains password in plaintext will be sent."
```
Comment (fballiano) [inline app/locale/en_US/Mage_Customer.csv:460]:
```suggestion
"Warning: email containing password in plaintext will be sent.","Warning: email containing password in plaintext will be sent."
```

**Comment by fballiano** [comment]
I've tested but I've found a few problems:
- when I set the password the first time (clicking the link in the email) it didn't work, the process finishes but I can't login with that password
- since I can't reuse that same link to re-set the password I tried to re-click the "email link to set password" checkbox from the admin and save the customer record, but I didn't get a second email

**Comment by kiatng** [comment]
@fballiano I tried to replicate your error:

[backend] Create a customer:
1. associate the website to admin;
2. Password Management > click _Email Link to Set Password_
3. Save

In the email: click on the set password link, and set new password. I was not able to login.

In backend, customer page:
1. F12 to bring out the developer panel
2. Remove disable attribute in the dropdown element _Associate to Website_
3. Change it to a website other than Admin.
4. Save.

Now go to frontend, I was able to login with the password I set.

I also repeat the above but this time I created a customer associated to a website other than Admin. I was able to login with the set password.

@fballiano What steps did you do?

**Comment by fballiano** [comment]
I've also retested a new customer creation
- backend, customers mask, create new customer
- i leave everything as it is (also the admin website)
- fill fistname, lastname, check the email link to set password
- I receive the email with the link
- click, set the password, save
- trying to login the password doesn't work :-(

**Comment by kiatng** [comment]
@fballiano Very strange, I retested, the email link worked for me,  I managed to received the email.
image

When creating a customer in backend with associated website set to admin, it is true that the password set in either admin or email will not work in any frontend store. But when the associated website is set to a frontend store, the password will work on that store. This behavior has nothing to do with this PR, it always been like this.

**Comment by fballiano** [comment]
true, if I select the website then the password I set clicking the like works.

but if I re-enter the newly created customer and re-check the "send password link" then I don't get another email :-\ maybe it's my setup, maybe mailhog is somehow filtering the double email, dunno

**Comment by fballiano** [comment]
if somebody can test this we could try to merge it although my tests failed

**Comment by kiatng** [comment]
@fballiano On the failure to login in frontend, is it because you associated the new customer to the default Admin website? Can you associate the customer to a frontend website? Then you should be able to login on that website with the password you set in backend.
image

**Comment by fballiano** [comment]
@ADDISON74 would you test this PR by @kiatng cause I have some problems and I'm not sure if it is only because of my environment :-)

**Comment by ADDISON74** [comment]
Indeed, there are logical issues related to this feature. Below you may find a simple test I did in the latest OpenMage without applying this PR.

If you create a `new customer account` in the Backend, the Associate to Website, Group, First Name, Last Name, Email and Password fields are mandatory. The password can be entered or auto-generated.

1 - No matter which password variant is chosen, after saving the customer, **he does not receive any email**, this is an big issue.

2 - If after saving the customer for the first time, one of the two password variant is used, surprisingly the customer receives an email with the followingcontent

screenshot-1

There is no doubt the email contains wrong information. The customer is not the one who set the password and he cannot change the password in his account, because he doesn't have the password to access it. 

My opinion is he must receive the password set up by the administrator even if it is in the plain-text format and for security reasons, the message he receives must contain a warning message and a link to set a new password ASAP.

**In conclusion**

- the issue of why no email is sent to the customer when the form is saved for the first time must be investigated.

- the email message must be changed by replacing the phrase "(the password you set when creating your account)" with the plain-text password. In addition,  a security warning and a link to reset the password ASAP is mandatory.

After fixing these two issues, we can evaluate whether this PR is still requested or not.

**Comment by fballiano** [comment]
I tried to retest but cannot make it work, I'll convert this PR to draft.

**Comment by kiatng** [comment]
There was a conflict, which is now resolved. Hopefully, this PR can now be tested.

**Comment by kiatng** [comment]
@ADDISON74 I think everything you said was said in PR #1205 

- The customer do not see the password in the email when an account is created by admin in backend
- The email message was wrong

This PR fixes the above issues by taking the suggestions from the comments in PR #1205. Ref the screenshots in PR #1283, 

- To not send plaintext password in email: when an account is created by admin in backend, add the ability to send an email to ask the customer to set new password 
- Add a new email template for the above

This PR is now so old, even I could not remember the details.

**Comment by kiatng** [comment]
@fballiano I edited the file LICENSE.txt directly in github by copying the entire text from [here](https://github.com/OpenMage/magento-lts/blob/main/LICENSE.txt). Git still thinks the file is not the same as base. I have also tried this many times in my local repo, even though the text is copied from base, so there is no diff, but git thinks otherwise. Git must be right, but I do not know how to make the text the same as base. Any idea?

**Comment by ADDISON74** [comment]
> @ADDISON74 I think everything you said was said in PR #1205
> 
> * The customer do not see the password in the email when an account is created by admin in backend
> * The email message was wrong
> 
> This PR fixes the above issues by taking the suggestions from the comments in PR #1205. Ref the screenshots in PR #1283,
> 
> * To not send plaintext password in email: when an account is created by admin in backend, add the ability to send an email to ask the customer to set new password
> * Add a new email template for the above
> 
> This PR is now so old, even I could not remember the details.

@kiatng - I carefully read everything that was said about this problem. in addition, I added a feedback testing the OpenMage version without applying this PR. I will check again this PR and let you know my feedback

**Comment by ADDISON74** [comment]
@kiatng - When you have some time please give me a sign to finish this PR. It increases security level in OM. Just read my comments.

**Comment by kiatng** [comment]
I will retest again as this has been a while.

**Comment by kiatng** [comment]
During retest, I discovered a bug. I fixed it and retested. It's good for me.

**Comment by ADDISON74** [comment]
Thank you for your time. in my case, I receive the message, visit the link and change the password, but I cannot log in with the new password. It could be an issue or not.

Do you receive a confirmation email after creating an account in Backend?

**Comment by kiatng** [comment]
I will retest on the can't-login after password reset when I have the time, hopefully within next couple of weeks.

File: app/code/core/Mage/Customer/Model/Customer.php [hunk 3]
```diff
@@ -880,6 +882,40 @@ public function sendPasswordResetConfirmationEmail()
...
+        $this->changeResetPasswordLinkToken($newResetPasswordLinkToken);
+
+        $template = $isNew
+            ? self::XML_PATH_PASSWORD_LINK_ACCOUNT_NEW_EMAIL_TEMPLATE
+            : self::XML_PATH_PASSWORD_LINK_EMAIL_TEMPLATE;
+
+        $this->_sendEmailTemplate(
+            $template,
+            self::XML_PATH_PASSWORD_LINK_EMAIL_IDENTITY,
+            array('customer' => $this),
```
Comment (sreichel) [inline app/code/core/Mage/Customer/Model/Customer.php:912]:
```suggestion
            ['customer' => $this],
```

File: app/code/core/Mage/Customer/Model/Customer.php [hunk 4]
```diff
@@ -880,6 +882,40 @@ public function sendPasswordResetConfirmationEmail()
         return $this;
     }
 
+    /**
+     * Send email with link to set password
+     *
+     * @bool $isNew Send welcome email?
+     * @return $this
+     */
+    public function sendPasswordLinkEmail($isNew = false)
```
Comment (sreichel) [inline app/code/core/Mage/Customer/Model/Customer.php:891]:
```suggestion
    public function sendPasswordLinkEmail(bool $isNew = false)
```

**Comment by ADDISON74** [comment]
@kiatng - I am able to log in, but please check it to confirm too. I did the test sending the link being logged in then logged out. In both cases it worked as expected. Also, I did not find any errors in logs.

We can go further and move it from Draft. I found issues related to this password changing but I will report each as issues.

**Comment by kiatng** [comment]
I retested password reset multiple times, each time I can login with the new password. So, I think it is ready for review.

**Review by ADDISON74 — APPROVED** [review]

**Review by sreichel — APPROVED** [review]

**Comment by colinmollenhour** [comment]
Nice work, but should this have gone in `next`? What is the impact for users that have customized templates stored in the database?

**Comment by sreichel** [comment]
Mhhh? No template changed. A new one has been added.

I can't see how it could break anything. What do i miss?
